src/hotspot/share/opto/library_call.cpp

```diff
@@ -6967,6 +6969,77 @@ bool LibraryCallKit::inline_base64_decodeBlock() {
   return true;
 }
 
+bool LibraryCallKit::inline_poly1305_processBlocks() {
+  address stubAddr;
+  const char *stubName;
+  assert(UsePolyIntrinsics, "need Poly intrinsics support");
+  assert(callee()->signature()->size() == 3, "poly1305_processBlocks has %d parameters", callee()->signature()->size());
+  stubAddr = StubRoutines::poly1305_processBlocks();
+  stubName = "poly1305_processBlocks";
+
+  if (!stubAddr) return false;
+  Node* polyObj = argument(0);
+  Node* input = argument(1);
+  Node* input_offset = argument(2);
+  Node* len = argument(3);
+
+  Node* accFace = load_field_from_object(polyObj, "a", "Lsun/security/util/math/MutableIntegerModuloP;");
+  assert(accFace != NULL, "Accumulator field is null");
+  const TypeInstPtr* ainst = _gvn.type(accFace)->isa_instptr();
+  assert(ainst != NULL, "Accumulator obj is null");
+  assert(ainst->is_loaded(), "MutableIntegerModuloP obj is not loaded");
+  ciKlass* klass_MutableElement = ainst->instance_klass()->find_klass(ciSymbol::make("sun/security/util/math/intpoly/IntegerPolynomial$MutableElement"));
+  assert(klass_MutableElement != NULL, "IntegerPolynomial$MutableElement class is null");
+  assert(klass_MutableElement->is_loaded(), "IntegerPolynomial$MutableElement class is not loaded");
+  ciInstanceKlass* instklass_MutableElement = klass_MutableElement->as_instance_klass();
+
+  const TypeKlassPtr* aklass = TypeKlassPtr::make(instklass_MutableElement);
+  const TypeOopPtr* atype = aklass->as_instance_type()->cast_to_ptr_type(TypePtr::NotNull);
+  Node* accObj = new CheckCastPPNode(control(), accFace, atype);
+  accObj = _gvn.transform(accObj);
+  Node* alimbs = load_field_from_object(accObj, "limbs", "[J");
+
+  Node* rFace = load_field_from_object(polyObj, "r", "Lsun/security/util/math/IntegerModuloP;"); //this.r.limbs
+  assert(rFace != NULL, "R field is null");
+  const TypeInstPtr* rinst = _gvn.type(rFace)->isa_instptr();
+  assert(rinst != NULL, "R obj is null");
+  assert(rinst->is_loaded(), "IntegerModuloP obj is not loaded");
+  ciKlass* klass_ImmutableElement = rinst->instance_klass()->find_klass(ciSymbol::make("sun/security/util/math/intpoly/IntegerPolynomial$ImmutableElement"));
+  assert(klass_ImmutableElement != NULL, "IntegerPolynomial$ImmutableElement class is null");
+  assert(klass_ImmutableElement->is_loaded(), "IntegerPolynomial$ImmutableElement class is not loaded");
+  ciInstanceKlass* instklass_ImmutableElement = klass_ImmutableElement->as_instance_klass();
+
+  const TypeKlassPtr* rklass = TypeKlassPtr::make(instklass_ImmutableElement);
+  const TypeOopPtr* rtype = rklass->as_instance_type()->cast_to_ptr_type(TypePtr::NotNull);
+  Node* rObj = new CheckCastPPNode(control(), rFace, rtype);
+  rObj = _gvn.transform(rObj);
+  Node* rlimbs = load_field_from_object(rObj, "limbs", "[J");
```

Comment:
@jnimeh if you could be particularly 'critical' here please? I generally know what I wanted to accomplish. And stepped through things with a debugger... but all the various IR types and conversions, I just don't know. I copied things from AES, which seem to work, as they do here, but I don't _understand_ the code.

i.e. recursive `getfield`s `((IntegerPolynomial$MutableElement)(this.a)).limbs` plus checks if we know field offsets: if (recursive) classes are loaded. But if not loaded, crashing with assert? Seems 'rude'. I think Poly1305 class constructor running would had forced the classes here to load so nothing to worry about, so I suppose assert is enough.)